storage/remote/write_handler.go

```diff
@@ -678,55 +607,108 @@ func hasDelta(md pmetric.Metrics) bool {
+	opts storage.AOptions,
+) (_ storage.SeriesRef, err error) {
+	if t > a.maxTime {
 		return 0, fmt.Errorf("%w: timestamp is too far in the future", storage.ErrOutOfBounds)
 	}
 
 	if h != nil && histogram.IsExponentialSchemaReserved(h.Schema) && h.Schema > histogram.ExponentialSchemaMax {
-		if err = h.ReduceResolution(histogram.ExponentialSchemaMax); err != nil {
+		if err := h.ReduceResolution(histogram.ExponentialSchemaMax); err != nil {
 			return 0, err
 		}
 	}
 	if fh != nil && histogram.IsExponentialSchemaReserved(fh.Schema) && fh.Schema > histogram.ExponentialSchemaMax {
-		if err = fh.ReduceResolution(histogram.ExponentialSchemaMax); err != nil {
+		if err := fh.ReduceResolution(histogram.ExponentialSchemaMax); err != nil {
 			return 0, err
 		}
 	}
 
-	if ref, err = app.Appender.AppendHistogram(ref, l, t, h, fh); err != nil {
-		return 0, err
+	retRef, err := a.AppenderV2.Append(ref, ls, st, t, v, h, fh, opts)
+
+	var sTyp string
+	switch {
+	case fh != nil, h != nil:
+		sTyp = sampleTypeHistogram
+		defer func() {
+			if err == nil {
+				a.stats.Histograms++
+			}
+		}()
+	default:
+		sTyp = sampleTypeFloat
+		defer func() {
+			if err == nil {
+				a.stats.Samples++
+			}
+		}()
+	}
+	var pErr *storage.AppendPartialError
+	if err != nil {
+		// Although Append does not currently return
+		// ErrDuplicateSampleForTimestamp there is
+		// a note indicating its inclusion in the future.
+		if errors.Is(err, storage.ErrOutOfOrderSample) ||
+			errors.Is(err, storage.ErrOutOfBounds) ||
+			errors.Is(err, storage.ErrDuplicateSampleForTimestamp) {
+			a.logger.Error("Error when appending sample", "err", err.Error(), "series", ls.String(), "timestamp", t, "sampleType", sTyp)
+		}
+
+		if !errors.As(err, &pErr) {
+			return retRef, err // Full append error.
+		}
 	}
-	return ref, nil
-}
 
-func (app *remoteWriteAppender) AppendExemplar(ref storage.SeriesRef, l labels.Labels, e exemplar.Exemplar) (storage.SeriesRef, error) {
-	if e.Ts > app.maxTime {
-		return 0, fmt.Errorf("%w: timestamp is too far in the future", storage.ErrOutOfBounds)
+	if ref != 0 {
+		// Non-zero input ref here means we had a successful append for this series (for the same exemplars).
+		// In this case, skip accounting for exemplar errors, if any. Exemplars are the same per the same series, so
+		// the same errors will be reported and duplicates will be silently skipped.
+		//
+		// This assumes exemplar errors are idempotent, which is a good enough assumption for this rare case.
+		// This is rare case as both OTLP and RW are optimized for a single sample per TS (although RW allows buffering).
+		return retRef, nil
 	}
 
-	ref, err := app.Appender.AppendExemplar(ref, l, e)
-	if err != nil {
-		return 0, err
+	if pErr != nil {
+		// Partial error due to exemplars, account for that.
+		a.stats.Exemplars += len(opts.Exemplars) - len(pErr.ExemplarErrors)
+		for _, e := range pErr.ExemplarErrors {
+			if errors.Is(err, storage.ErrOutOfOrderExemplar) {
```

Comment:
This looks like a bug to me.
```suggestion
			if errors.Is(e, storage.ErrOutOfOrderExemplar) {
```